R/utilities-sensitivity-analysis.R

```diff
@@ -403,30 +386,78 @@ runPopulationSensitivityAnalysis <- function(structureSet, settings, logFolder =
 }
 
 
+#' @title getPKResultsDataFrame
+#' @description Read PK parameter results into a dataframe and set QuantityPath,Parameter and Unit columns as factors
+#' @param structureSet `SimulationStructure` R6 class object
+#' @return pkResultsDataFrame, a dataframe storing the contents of the CSV file with path pkParameterResultsFilePath
+#' @import ospsuite
+getPKResultsDataFrame <- function(structureSet) {
+  pkResultsDataFrame <- read.csv(structureSet$pkAnalysisResultsFileNames,
+    encoding = "UTF-8",
+    check.names = FALSE,
+    stringsAsFactors = FALSE
+  )
+
+  # get a list pkParameterNamesEachOutput where field names are the paths to the output andthe names of the pkParameters for each output in the current simulation set
+  pkParameterNamesEachOutput <- list()
+
+  for (n in seq_along(structureSet$simulationSet$outputs)) {
+    op <- structureSet$simulationSet$outputs[[n]]
+    if (!is.null(op$pkParameters)) {
+      # case where pkParameters are specified for each output path
+      pkParameterNamesEachOutput[[op$path]] <- unname(sapply(
+        op$pkParameters,
+        function(y) {
+          y$pkParameter
+        }
+      ))
+    } else {
+      # case where no pkParameters are specified for an output and SA performed for all pk parameters by default
+      pkResultsDataFrameThisOutput <- pkResultsDataFrame[ pkResultsDataFrame$QuantityPath == op$path, ]
+      pkParameterNamesEachOutput[[op$path]] <- unique(pkResultsDataFrameThisOutput$Parameter)
+    }
+  }

```

Comment:
You can simplify the loop as below:
`for (output in structureSet$simulationSet$outputs){ if(!is.null(output$pkParameters)) ... }`

Besides, for character type variables, I think `pkResultsDataFrame$QuantityPath %in% op$path` is better than `pkResultsDataFrame$QuantityPath == op$path`

Would it also be better to start the loop with the `else` content and then only use an `if` to filter and check that the user selected pkParameters are in pkResultsDataFrame ?

Reply:
Fixed